### PR TITLE
Fix variable type inference precedence

### LINKED ISSUE #18689: Completion type inference infers from the first declaration in a document rather than the closest
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

```powershell
$script = 'function Test-Function {
    [hashtable] $myVar = @{ Completion = ''error'' }
}


[datetime] $myVar = Get-Date
$myVar.'

TabExpansion2 -inputScript $script -cursorColumn $script.Length | % CompletionMatches
```

### Expected behavior

```console
CompletionText        ListItemText         ResultType ToolTip
--------------        ------------         ---------- -------
Date                  Date                   Property datetime Date {get;}
Day                   Day                    Property int Day {get;}
DayOfWeek             DayOfWeek              Property System.DayOfWeek DayOfWeek {get;}
DayOfYear             DayOfYear              Property int DayOfYear {get;}
Hour                  Hour                   Property int Hour {get;}
Kind                  Kind                   Property System.DateTimeKind Kind {get;}
Microsecond           Microsecond            Property int Microsecond {get;}
Millisecond           Millisecond            Property int Millisecond {get;}
Minute                Minute                 Property int Minute {get;}
Month                 Month                  Property int Month {get;}
Nanosecond            Nanosecond             Property int Nanosecond {get;}
Second                Second                 Property int Second {get;}
Ticks                 Ticks                  Property long Ticks {get;}
TimeOfDay             TimeOfDay              Property timespan TimeOfDay {get;}
Year                  Year                   Property int Year {get;}
(etc)
```

### Actual behavior

```console
CompletionText     ListItemText      ResultType ToolTip
--------------     ------------      ---------- -------
Count              Count               Property int Count { get; }
IsFixedSize        IsFixedSize         Property bool IsFixedSize { get; }
IsReadOnly         IsReadOnly          Property bool IsReadOnly { get; }
IsSynchronized     IsSynchronized      Property bool IsSynchronized { get; }
Keys               Keys                Property System.Collections.ICollection Keys { get; }
SyncRoot           SyncRoot            Property System.Object SyncRoot { get; }
Values             Values              Property System.Collections.ICollection Values { get; }
Add(               Add                   Method void Add(System.Object key, System.Object value)…
Clear(             Clear                 Method void Clear()…
Clone(             Clone                 Method System.Object Clone()…
Contains(          Contains              Method bool Contains(System.Object key)…
ContainsKey(       ContainsKey           Method bool ContainsKey(System.Object key)
ContainsValue(     ContainsValue         Method bool ContainsValue(System.Object value)
CopyTo(            CopyTo                Method void CopyTo(array array, int arrayIndex)…
Equals(            Equals                Method bool Equals(System.Object obj)
GetEnumerator(     GetEnumerator         Method System.Collections.IDictionaryEnumerator GetEnumerator()…
GetHashCode(       GetHashCode           Method int GetHashCode()
GetObjectData(     GetObjectData         Method void GetObjectData(System.Runtime.Serialization.SerializationInfo info, System.Runtime.Ser…
GetType(           GetType               Method type GetType()
OnDeserialization( OnDeserialization     Method void OnDeserialization(System.Object sender)…
Remove(            Remove                Method void Remove(System.Object key)…
ToString(          ToString              Method string ToString()
```

### Error details

_No response_

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.3.0
PSEdition                      Core
GitCommitId                    7.3.0
OS                             Microsoft Windows 10.0.22621
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by MartinGC94**
# PR Summary
Fixes https://github.com/PowerShell/PowerShell/issues/18689
So instead of using the first variable assignment before the cursor to infer the type, it instead takes the last variable assignment before the cursor.  
Removed a condition that could never be true because the loop before it always ensures that the parent property is null.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by AndrewJoncas** [comment]
Doesn't this approach of reversing the break this pattern of coding common to ps1 scripts?

'''Powershell
$script = '[CmdletBinding()]

param (

    # PS1 Script Parameter

    [Parameter()]

    [hashtable]

    $ParameterName

)

 

function Verb-Noun {

    [CmdletBinding()]

    param (

        # PS1 Function Parameter

        [Parameter()]

        [string]

        $ParameterName

    )

   

}

 

$ParameterName.'

 

TabExpansion2 -inputScript $script -cursorColumn $script.Length | select -expandproperty CompletionMatches

'''

**Comment by MartinGC94** [comment]
So, apparently VisitAssignmentStatement returns whatever is on the left: https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs#L977  
This caused some issues in my first test `[string]$TestVar = "";[hashtable]$TestVar = @{};$TestVar` where the typeinference saw it like the ast returned 3 types (string, hashtable x2) which was obviously not correct. If I fix that so AssignmentStatements return nothing then it's incorrect when used inside parentheses like this: `$Var1 = ($Var2 = ls)` because in that case the AssignmentStatement does return something.

@SeeminglyScience What's your thoughts on this? How do I fix this properly?

**Comment by MartinGC94** [comment]
> Doesn't this approach of reversing the order break this pattern of coding common to ps1 scripts?

Not this exact scenario because parameters are handled before variable assignments, so it will see the hashtable parameter before the other function parameter. But you are right that the approach isn't perfect. If you define a variable, and then a function like this:

```
[hashtable]$ParameterName=@{}
function Verb-Noun {
    [CmdletBinding()]
    param
    (
        [Parameter()]
        [string]
        $ParameterName
    )
}
$ParameterName.<Tab>
```
then it will see the function parameter instead of the script variable. I will need to look into that.

**Comment by SeeminglyScience** [comment]
Mmmm tricky! I'm inclined to think that it should continue to infer the type on the RHS, but *only* of the closest assignment.

But then if you do

```powershell
[string] $thing = 'something'
$thing = 10
$thing.<tab>
```

then you'll get `int`. And maybe that's fine but it sure doesn't feel good. The *ideal* thing would be if it looped through parents and looked through assignments to find the closest "declaration" but *not recurse* into the parent's children. That way you could avoid `& { [string] $a = 10 }; $a = gci; $a.<tab>` finding `string`.

Though even that's not necessarily perfect because then you'd miss declarations in a foreach loop. BUT you could argue that you should have the declaration outside of the loop, and it's fine to miss it in this case.

Sorry I got kind of ramble-y there, but does that make sense? Not in a "just go do that" kind of way, I'm still thinking over exactly what should be done, just kind of talking it out.

**Comment by MartinGC94** [comment]
I found a simple solution, just add a special case for AssignmentStatements inside ParenExpressions. On a related note, do we have an official name for this: `$Var1 = ($Var2 = ls)`. People on the internet call it "variable squeezing" but I haven't seen any official documents with this name.

**Comment by SeeminglyScience** [comment]
> I found a simple solution, just add a special case for AssignmentStatements inside ParenExpressions. On a related note, do we have an official name for this: `$Var1 = ($Var2 = ls)`. People on the internet call it "variable squeezing" but I haven't seen any official documents with this name.

So `Paren` can force it in scenarios where it doesn't typically happen, but it's implicit with `$var1 = $var2 = ls`. I'm inclined to say the best solution would be to *only* look at the closest assignment prior to itself if that makes sense

**Comment by MartinGC94** [comment]
I've reverted the changes related to the assignmentstatementast type inference. I will (probably) look into fixing it in the future but it doesn't belong in this PR.

File: test/powershell/engine/Api/TypeInference.Tests.ps1 [hunk 1]
```diff
@@ -1367,6 +1367,24 @@ Describe "Type inference Tests" -tags "CI" {
         $res.Name -join ' ' | Should -Be "System.IO.FileInfo System.IO.DirectoryInfo"
     }
 
+    It 'Infers closest variable type' {
+        $res = [AstTypeInference]::InferTypeOf( { [string]$TestVar = "";[hashtable]$TestVar = @{};$TestVar }.Ast)
```
Comment (JamesWTruher) [inline test/powershell/engine/Api/TypeInference.Tests.ps1:1371]: should there be some sort of order checking here?

Comment (MartinGC94) [reply test/powershell/engine/Api/TypeInference.Tests.ps1:1371]: No. There shouldn't even be multiple inferred types but because assignments are included when inferring the output from a scriptblock I need that `Select -Last 1` statement.  
I tried fixing the issue earlier but it's tricky because sometimes assignments actually do output something.

File: src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs [hunk 2]
```diff
@@ -2490,7 +2485,8 @@ public static bool AstAssignsToSameVariable(this VariableExpressionAst variableA
             if (parameterAst != null)
             {
                 return variableAstVariablePath.IsUnscopedVariable &&
-                       parameterAst.Name.VariablePath.UnqualifiedPath.Equals(variableAstVariablePath.UnqualifiedPath, StringComparison.OrdinalIgnoreCase);
+                       parameterAst.Name.VariablePath.UnqualifiedPath.Equals(variableAstVariablePath.UnqualifiedPath, StringComparison.OrdinalIgnoreCase) &&
+                       parameterAst.Parent.Parent.Extent.EndOffset > variableAst.Extent.StartOffset;
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:2489]: @MartinGC94 This change came from the commit https://github.com/PowerShell/PowerShell/pull/18691/commits/c91e29f27ef323a52429ae59661692001ddd8348, which was reverted by https://github.com/PowerShell/PowerShell/pull/18691/commits/1289c4e64cdb5de26bce325d36968b771d2bbd5a. However, this change was left behind. Do you mean to revert this change as well?

Comment (MartinGC94) [reply src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs:2489]: No, that was a fix for this: https://github.com/PowerShell/PowerShell/pull/18691#issuecomment-1332909496 I was just lazy and put 2 changes in 1 commit which was clearly a bad idea 😄

**Review by daxian-dbw — APPROVED** [review]
LGTM
